Blocks: Support reusable nested blocks (reusable blocks refactor)

aduth opened the pull request:
Supersedes #5010
Cherry-picks 840fbec793462039b3cd3279ae807516a0a1a27c and ee61e18cc9c846f032ba663c615b5f9716eac128 from #5010 
Cherry-picks dde17871bd7a57956c0ada10f1df7a8fd7dda117 and 80e03deba7d7737099eed1cdd02e6cc352792c25 from #5223
Cherry-picks 66fdda52ec7177c1a8154e2df2a75da852c813ca from #5220

This pull request seeks to support reusable nested blocks. Unlike #5010, the changes here do not expand the reusable blocks API to support nested blocks, but rather refactor reusable blocks to serve as pointers to the original block references, therefore having better natural support for nested blocks.

A few of the changes have been brought along from #5010:

1. Move context assignment of inner block list from `BlockListBlock` to `BlockEdit`, as the latter is the component used in rendering from `core/block`'s `edit`
   - Downside is a circular dependency from `blocks` to `editor`, as `BlockEdit` requires access to the `BlockList` component
2. Update reusable block UI behaviors to not abort editing when the block becomes deselected, as this is incompatible with nested blocks where selection changes to the inner block

A complication of reusable blocks as pointers it that we need to add support for the blocks state to receive arbitrary blocks after initial load without dirtying state (the loaded result of a reusable block should be merged into but not dirtying to blocks state). New options of `ignoreTypes` have been added to both the `withChangeDetection` and `withHistory` higher-order reducers to accommodate this.

__Testing instructions:__

Verify there are no regressions in the behavior of reusable blocks.

Ensure that you can save a Columns block, including content, as a reusable block.

Ensure unit tests pass:

```
npm test
```

__Remaining tasks:__

- [x] Reusable blocks effects tests

aduth committed:
Blocks: Avoid dismissing editable controls

In nested context, block selection changes from reusable block to the inner block being edited, but we want to keep the UI shown

aduth committed:
Block: Move block context provides into BlockEdit

aduth committed:
Revert "Blocks: Disable Convert to Reusable for nested blocks"

This reverts commit 887205810487cf987b2f639a9bc18b57ae093765.

aduth committed:
State: Enhance history reducer utils to support ignores

aduth committed:
Block List: Hide disabled default block appender

e.g. within Disabled component, e.g. within reusable nested block

aduth committed:
State: Update reusable blocks as pointers to state blocks

aduth committed:
Block List: Inject inner block list creator as function

Generate BlockList from block's own utils, which has advantage of:

- Not having circular dependency from blocks to editor
- Respecting block menu and contextual toolbar props

aduth commented:
Rebased to resolve conflicts. Also made an update in 5c6cd14 which avoids referencing the `wp.blocks.BlockList` global and respects contextual toolbar and block menu settings.

For the editable controls removed in 7d0bdde, I think there might be a solution involving [`withFocusOutside`](https://github.com/WordPress/gutenberg/tree/master/components/higher-order/with-focus-outside), but it's a bit tricky because focus is lost when transitioning into Edit mode because the Edit button disappears, so we need to find some other way to transition focus back into the block after editing has started. We _should_ be able to just dispatch `selectBlock` on the block and rely on [the selection `focusTabbable` behavior](https://github.com/WordPress/gutenberg/blob/f2bf8a1c66eb80cbf30c29c1cc00b08ad98bd9df/editor/components/block-list/block.js#L177-L179), but this was not working in my testing (on a Columns reusable block).

noisysocks commented on editor/store/test/selectors.js:
```diff
@@ -2477,25 +2498,50 @@ describe( 'selectors', () => {
 	} );
 
+	describe( 'isFetchingReusableBlock', () => {
+		it( 'should return false when the block is not being saved', () => {
```
s/saved/fetched/

noisysocks replied:
Fixed in c348a43763761abf545b63d80664f248f93b435b.

noisysocks commented on editor/store/test/selectors.js:
```diff
@@ -2477,25 +2498,50 @@ describe( 'selectors', () => {
+			expect( isFetching ).toBe( false );
+		} );
+
+		it( 'should return true when the block is being saved', () => {
```
s/saved/fetched/

noisysocks replied:
Fixed in c348a43763761abf545b63d80664f248f93b435b.

noisysocks reviewed (dismissed):
This is looking great. Really nice work 😍 

Two issues I noticed in my testing:

1. Reusable blocks are not displaying at all in the inserter. They display correctly on `master`. I can see that the `FETCH` happens, so my guess is that it's a regression in the `getInserterItems` selector.
2. If you have a reusable block in the editor and you reload the page, there is a moment where the _Block has been deleted or is unavailable_ message appears for about a single frame, right after the loading spinner hides. My guess is that this message appears in between when `FETCH_REUSABLE_BLOCKS_SUCCESS` is dispatched and when `RECEIVE_REUSABLE_BLOCKS` is dispatched.

pento commented:
🎶_Ain't no party like a recursive block party, because a recursive block party don't stop._ 🎶

I ran into the same problem with the inserter as @noisysocks, so I did this by manually editing the shared block in the database.

- Create a Nested Block
- Save it as a Shared Block
- Edit the new Shared block manually, to include itself as a column.
- Reload Gutenberg
- 🎉

It'd probably be a good idea to implement recursion limits on this. I'd probably start with not allowing a nested shared block to contain itself, then let's see if there's potential for abuse of nested block trees before we worry too much about putting limits on it.

noisysocks committed:
Typo: saved -> fetched

noisysocks committed:
Typo: preferencially -> preferentially

noisysocks committed:
Explain why dispatching RECEIVE_BLOCKS is necessary

noisysocks committed:
Remove unnecessary getState variable

noisysocks committed:
Update getInserterItems to respect new reusable block data layout

Makes `getInserterItems` and `getFrecentInserterItems` respect that
reusable blocks now point to a block that is elsewhere in the editor
state. This makes reusable blocks again appear in the inserter.

noisysocks committed:
Provide createInnerBlockList context in BlockPreview

Allow reusable nested blocks to be previewed in the inserter by having
BlockPreview inject a createInnerBlockList function via context.

noisysocks committed:
Fix flash of 'Not found' error message when fetching a reusable block

Re-order the FETCH_REUSABLE_BLOCKS effect so that the reusable block is
added to the store before it is marked as no longer being fetched. This
prevents core/block from briefly flashing the 'Not found' error message
in between dispatches.

noisysocks committed:
When converting reusable -> regular, replace the block with a copy

Making a copy of the referenced block prevents the regular block from
being removed should the reuasble block be later deleted.

noisysocks commented on blocks/library/block/index.js:
```diff
@@ -40,120 +39,126 @@ class ReusableBlockEdit extends Component {
+				{ element }
+				{ ( isSelected || isEditing ) && (
+					<ReusableBlockEditPanel
+						key="panel"
```
No longer relevant.

noisysocks commented on blocks/block-edit/index.js:
```diff
@@ -2,51 +2,90 @@
+
+		// `edit` and `save` are functions or components describing the markup
+		// with which a block is displayed. If `blockType` is valid, assign
+		// them preferencially as the render value for the block.
```
Fixed in f4eb39134ae59698b58e56c3cba70d07bd8f250b.

noisysocks commented on blocks/block-edit/index.js:
```diff
@@ -2,51 +2,90 @@
+	} ) ),
+	withAPIData( ( { postType } ) => ( {
+		user: `/wp/v2/users/me?post_type=${ postType }&context=edit`,
+	} ) ),
```
Probably not necessary for now.

aduth replied:
With #5219, I might imagine this could be a simple selector on a core data module.

noisysocks commented on editor/store/effects.js:
```diff
@@ -355,9 +364,11 @@ export default {
 
 		const { id } = action;
 		const { getState, dispatch } = store;
+		const state = getState();
```
Fixed in 4e1026f5350ac28d825c2387ee7ce6859bfb18f5.

noisysocks commented on editor/store/effects.js:
```diff
@@ -439,25 +455,38 @@ export default {
+			} )
+		) );
+
+		dispatch( receiveBlocks( [ parsedBlock ] ) );
```
Fixed in 08333d32f3ca01599b6bc4eeae9e4cc26ee76b33.

noisysocks commented:
I rebased, addressed my own feedback, fixed those two issues that I described, and fixed an additional issue I discovered where deleting a reusable block would cause blocks that were created via 'Convert to Regular Block' to be removed.

Still to come: update the effects tests that were removed.

aduth commented on editor/store/effects.js:
```diff
@@ -439,25 +455,38 @@ export default {
+		const parsedBlock = getBlock( getState(), action.uid );
+		const reusableBlock = {
+			id: uniqueId( 'reusable' ),
+			uid: parsedBlock.uid,
```
Couldn't naming the property `blockUID` achieve the same effect as what you're describing?

noisysocks commented:
This should be good to go. Care to give it a quick check, @aduth?

aduth commented on editor/components/block-preview/index.js:
```diff
@@ -6,11 +6,13 @@ import { noop } from 'lodash';
 /**
  * Internal dependencies
  */
+import { createInnerBlockList } from '../block-list/utils';
```
I don't love that we reach into the local utilities of a separate component. Seems like if we need it in both places, might be best to extract up a level to a common place.

noisysocks replied:
👌 moved in e99f333aaeb2626b40f4c03c6fb8d5996443d50c.

aduth commented on editor/store/selectors.js:
```diff
@@ -1277,7 +1312,7 @@ export function isSavingReusableBlock( state, ref ) {
  */
 export function getReusableBlocks( state ) {
-	return Object.values( state.reusableBlocks.data );
+	return Object.keys( state.reusableBlocks.data ).map( ( ref ) => getReusableBlock( state, ref ) );
```
Minor: Lodash's `_.map` works well on objects for this purpose (also trivially more performant by avoiding a second iteration):

```js
return map(
	state.reusableBlocks.data,
	( data, ref ) => getReusableBlock( state, ref )
);
```

noisysocks replied:
👌 changed in 7d87636ec2d95a9e678423fa66390ee2649af4ad.

aduth reviewed (commented):
I'm noticing that the default block appender no longer shows at all when editing nested blocks. Thinking this could perhaps be related to recent refactoring with it (#5478).

We should probably also wait to merge this until after the next release, since it's large enough to risk breakage.

noisysocks committed:
Avoid iterating through reusable blocks twice

Use _.map to avoid iterating through the reusable blocks twice in
`getReusableBlocks`.

noisysocks committed:
Move createInnerBlockList into `editor/utils`

noisysocks commented:
> I'm noticing that the default block appender no longer shows at all when editing nested blocks. 

It looks like this is broken in `master`. I ran `git bisect` and it seems that 530a66e0cc168ffc153da89d4c4f03c6e75fca1b is what caused it.

> We should probably also wait to merge this until after the next release, since it's large enough to risk breakage.

👍

aduth commented:
This is looking good, and I'd like it to have a few days to sit in master ahead of the next release 👍 

The regression with columns default appender should be addressed separately, since it was not introduced here.

aduth commented:
Thanks for your help with this pull request @noisysocks
